### PR TITLE
Bug 1876515: Fix inline taskSpec error in pipeline page

### PR DESCRIPTION
**Pull request by karthikjeeyar**
Fixes: https://issues.redhat.com/browse/ODC-4724

Problem:
Issue with inline spec over taskref causes the pipeline details page to break

Solution:
Handle the inline spec issue and skip the visualization

Screenshots:
image

Test cases:
image
PipelineVisualization tests
image

pipeline-actions tests
image

### CODE REVIEW

File: frontend/packages/dev-console/src/utils/pipeline-utils.ts [hunk 1]
```diff
@@ -181,7 +181,16 @@ const appendPipelineRunStatus = (pipeline, pipelineRun) => {
     return mTask;
   });
 };
-
+export const hasInlineTaskSpec = (pipeline: K8sResourceKind) => {
+  const tasks = (pipeline.spec && pipeline.spec.tasks) || [];
```
Comment (divyanshiGupta) [inline frontend/packages/dev-console/src/utils/pipeline-utils.ts:185]:
```suggestion
  const tasks = pipeline.spec?.tasks ?? [];
```

File: frontend/packages/dev-console/src/utils/pipeline-utils.ts [hunk 2]
```diff
@@ -192,7 +201,7 @@ export const getPipelineTasks = (
 ): PipelineVisualizationTaskItem[][] => {
   // Each unit in 'out' array is termed as stage | out = [stage1 = [task1], stage2 = [task2,task3], stage3 = [task4]]
   const out = [];
-  if (!pipeline.spec || !pipeline.spec.tasks) {
+  if (!pipeline.spec || !pipeline.spec.tasks || hasInlineTaskSpec(pipeline)) {
```
Comment (divyanshiGupta) [inline frontend/packages/dev-console/src/utils/pipeline-utils.ts:204]:
```suggestion
  if (!pipeline.spec?.tasks || hasInlineTaskSpec(pipeline)) {
```

**Review by divyanshiGupta — CHANGES_REQUESTED** [review]

**Review by sahil143 — APPROVED** [review]
/lgtm

verified changes locally

**Comment by sahil143** [comment]
/lgtm cancel

It breaks in the Edit Pipeline page
Screenshot 2020-09-09 at 3 26 47 PM

**Comment by karthikjeeyar** [comment]
Added the check in edit scenario, so that if there is an inline taskSpec, it will not throw error and will allow us to add/modify the tasks in the builder form.
image

**Comment by andrewballantyne** [comment]
@karthikjeeyar I think we need UX here... I don't think it's appropriate to edit a Pipeline and clear out the tasks because they are using inline spec - this could have wildly unexpected results for the user. I think we need to take a path of least resistance here - it is an unsupported feature of the DevConsole. Maybe we showcase a message saying "Using tasks inlineSpec, which is unsupported" and removing the Edit Pipeline kebab option. 

cc @bgliwa01 / @openshift/team-devconsole-ux

File: frontend/packages/dev-console/src/utils/pipeline-utils.ts [hunk 3]
```diff
@@ -181,7 +181,16 @@ const appendPipelineRunStatus = (pipeline, pipelineRun) => {
...
 };
-
+export const hasInlineTaskSpec = (pipeline: K8sResourceKind) => {
+  const tasks = pipeline?.spec.tasks ?? [];
+  let hasInlineSpec = false;
+  tasks.forEach((task) => {
+    if (!hasInlineSpec) {
+      hasInlineSpec = !!(task.taskSpec && !task.taskRef);
+    }
+  });
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/pipeline-utils.ts:191]: You're re-implementing `Array.prototype.some`... best just [use it instead](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some).

Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/utils/pipeline-utils.ts]: Updated it.

**Comment by bgliwa01** [comment]
@karthikjeeyar I agree with @andrewballantyne we should remove the edit pipeline option in the kebab dropdown and then we can display the message "This Pipeline cannot be visualized. Pipeline taskSpec is not supported."

**Comment by andrewballantyne** [comment]
> @karthikjeeyar I agree with @andrewballantyne we should remove the edit pipeline option in the kebab dropdown and then we can display the message "This Pipeline cannot be visualized. Pipeline taskSpec is not supported."

@karthikjeeyar When you do this, make sure there is a gap under the message before the first label.

Thanks @bgliwa01 !

**Comment by karthikjeeyar** [comment]
Updated the code to hide the edit pipeline option and added the UX provided message to the visualization area.
 @bgliwa01 Please take a look at the updated screenshots.
cc: @andrewballantyne

File: frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualization.tsx [hunk 4]
```diff
@@ -17,21 +18,32 @@ const PipelineVisualization: React.FC<PipelineTopologyVisualizationProps> = ({
...
+      <Alert
+        variant="info"
+        isInline
+        title="This Pipeline cannot be visualized. Pipeline taskSpec is not supported."
+      />
+    );
+  } else if (nodes.length === 0 && edges.length === 0) {
     // Nothing to render
     // TODO: Confirm wording with UX; ODC-1860
-    return <Alert variant="info" isInline title="This Pipeline has no tasks to visualize." />;
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualization.tsx:24]: Why did you remove the return? I think we can just have another return here.

File: frontend/packages/dev-console/src/utils/pipeline-utils.ts [hunk 5]
```diff
@@ -181,7 +181,10 @@ const appendPipelineRunStatus = (pipeline, pipelineRun) => {
     return mTask;
   });
 };
-
+export const hasInlineTaskSpec = (pipeline: K8sResourceKind) => {
+  const tasks = pipeline?.spec.tasks ?? [];
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/pipeline-utils.ts:185]: `spec` is not optional.

File: frontend/packages/dev-console/src/utils/pipeline-utils.ts [hunk 6]
```diff
@@ -192,7 +195,7 @@ export const getPipelineTasks = (
 ): PipelineVisualizationTaskItem[][] => {
   // Each unit in 'out' array is termed as stage | out = [stage1 = [task1], stage2 = [task2,task3], stage3 = [task4]]
   const out = [];
-  if (!pipeline.spec || !pipeline.spec.tasks) {
+  if (!pipeline?.spec.tasks) {
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/pipeline-utils.ts:198]: Don't think this can ever happen anymore. Tasks of 1 is required by the operator. We can keep this if you change it to look for tasks also being an empty array. 

Also `spec` is not optional.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/packages/dev-console/src/utils/__tests__/pipeline-test-data.ts [hunk 7]
```diff
@@ -12,7 +12,7 @@ export const mockPipelinesJSON: K8sResourceKind[] = [
       params: [
         {
           name: 'APP_NAME',
-          type: 'string',
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/__tests__/pipeline-test-data.ts:15]: `type` should be allowed 🤔  Why did you remove this?

Comment (andrewballantyne) [reply frontend/packages/dev-console/src/utils/__tests__/pipeline-test-data.ts:15]: Ah because it's not part of the type 😞 

Can you just add it to the type instead? 
It's [supported by upstream](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#specifying-parameters)
> Each declared parameter has a type field, which can be set to either array or string.

We'll want the type to probably be:

```
  type: 'string' | 'array';
```
Instead of a straight `string` type... this way we can easily see what we support.

**Review by bgliwa01 — APPROVED** [review]
lgtm

File: frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualization.tsx (hunk 4 above)
Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualization.tsx:24]: I have moved it inside the block, so that it gets the margin-bottom style (for the gap underneath the message) from the `odc-pipeline-visualization` class.

Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualization.tsx:24]: Ah, I see... why don't we clean this up more and instead of treating errors differently, just have a variable called `content` or something and have 3 if-else statements... landing on rendering the Component as the fallback `else`.

It'll look cleaner and be easier to read.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualization.tsx (hunk 4 above)
Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualization.tsx:24]: I agree, its more readable now. Updated it

**Review by andrewballantyne — APPROVED** [review]

File: frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.tsx [hunk 8]
```diff
@@ -33,10 +33,8 @@ interface PipelineVisualizationTaskProp {
   namespace: string;
   task: {
     name?: string;
-    taskRef: {
-      name: string;
-      kind?: string;
-    };
+    taskSpec?: PipelineTaskSpec;
+    taskRef?: PipelineTaskRef;
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.tsx:37]: :+1:

**Review by jerolimov — APPROVED** [review]
Tested it locally, works fine :+1: 

/lgtm
